### PR TITLE
[57] ADR |  Use Docker for running the application

### LINKED ISSUE #57: ADR: Use Docker for running the app
Write an ADR to reflect our decision to use Docker for running the portal client both locally during development, and in deployed environments.

### PR DESCRIPTION
**Pull request by esacteksab**
## Description 

An ADR on using Docker for local development and a Docker image as a deployment artifact.

Fixes #57 

## Review Notes

A WIP. I'd like @suzubara to add anything. Please let me know if I missed something or your thoughts in general.

### CODE REVIEW

**Comment by suzubara** [comment]
@esacteksab This looks great to me, I think you covered all of the points I was thinking of!

**Review by suzubara — APPROVED** [review]

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+### Production Deployment
+
+- Ship a Docker contianer as the artifact to be deployed.
+- Production server clones local repo and builds locally on production host.
+- Ship zip or tarball to remote host, extract and run locally on remote production host.
+
+## Decision Outcome
+
+Development done locally and application (and supporting services) running inside a Docker container. As we as shipping a Docker container as the artifact to be deployed to production. After some discussion, we have decided on these options for the following reasons:
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:32]: copy edit only: missing word or typo in beginning of second sentence?

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+- A git clone may pull down more than what's necessary to run the production application resulting in potential vulnerability exploits as well as disk usage.
+- Potential for being rate-limited by VCS host or CDN.
+- Larger attack surface than a Docker container based on a distroless image.
+- Build time is multipled by number of instances leading to longer deployments than a singular artifact (Docker image).
+
+### Ship zip or tarball to remote host, extract and run locally on remote production host
+
+#### Pros
+
+- It's a well established pattern in the DevOps community with plenty of working examplest to copy from.
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:137]: smol typo on 'examples'

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+### Local Development
+
+- Development done on local workstation and application (and supporting services) running on local workstation
+- Development done inside Docker container and application (and supporting services) running inside Docker container.
+- Development done locally and application (and supporting services) running inside a Docker container.
+
+### Production Deployment
+
+- Ship a Docker contianer as the artifact to be deployed.
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:26]: contianer -> container

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+- Managing mutliple applications and services is more effort than `docker compose up`.
+- Possibility of mismatched versions between environments.
+
+### Development done inside Docker container and application (and supporting services) running inside Docker container
+
+#### Pros
+
+- Nice and tidy. _Everything_ is in the container.
+- As long as everyone is running the same version of container, everyone has the same environment so should have the same experiences.
+- No need to install any applications locally keeping machine more pristine or as pristene as OS X with Docker installed can be.
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:74]: pristene -> pristine

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+- Only the application and the necessary dependencies exist in the container, creating a smaller attack surface.
+- Docker container can be mounted with a read-only file system allowing for greater security.
+- Deployment and running of a Docker container is far simpler than the other methods mentioned below.
+- We can inject secure environment variables at runtime allowing for greater security.
+- The same image that's used for local development, verification, testing and validation in dev and test is the same image that goes to production ensuring consistency.
+
+#### Cons
+
+- Docker and it's ism's
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:114]: it's -> its

File: docs/adr/0005-use-Docker-for-running-the-application.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+#### Pros
+
+- It's a well established pattern in the DevOps community with plenty of working examples to copy from.
+
+#### Cons
+
+- Assumes 100% uptime of VCS repository.
+- Assumes remote host is capable of reaching VCS repository.
+- Requires additional software (git) and it's dependencies to be installed on a production server.
```
Comment (abbyoung) [inline docs/adr/0005-use-Docker-for-running-the-application.md:127]: it's -> its

**Review by abbyoung — APPROVED** [review]
comprehensive, excellent! some copy edits but content is 💯

**Comment by esacteksab** [comment]
> comprehensive, excellent! some copy edits but content is 💯

I'm not worthy!

**Pull request merged**